tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+  HloInstruction* transpose;
+  HloInstruction* input;
+  HloInstruction* constant;
+  if (!Match(lhs,
+             m::Reshape(&reshape, m::Transpose(&transpose, m::Op(&input)))) ||
+      !Match(rhs, m::Constant(&constant))) {
+    return nullptr;
+  }
+
+  // Check if reshape squishes some dims into one dim, and that if this one
+  // dim is the dot's lhs contracting dim.
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
+    return nullptr;
+  }
+  for (const auto &pair : unmodified_dims) {
+    if (pair.second == lhs_contracting_dims[0]) {
+      // If the contracting dim is unmodified, and only one dim is modified,
+      // then that dim must be a non-contracting dim.
+      return nullptr;
+    }
+  }
+  // Virtually pull the reshape into the dot. Now the dot is equivalent to a
+  // new dot with "unsquished" lhs contracting dims. We don't need to actually
+  // create a new dot instruction. We can just keep track of lhs and
+  // lhs_contracting_dims.
+  CHECK(lhs->operand(0)->shape().rank() > lhs->shape().rank());
+  for (int64 i = 1; i <= lhs->operand(0)->shape().rank() - lhs->shape().rank();
+       ++i) {
+    lhs_contracting_dims.Add(lhs_contracting_dims[0] + i);
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // Check if lhs_transpose only permutes the contracting dims.
+  const auto& transpose_dims = lhs->dimensions();
+  for (int64 i = 0; i < transpose_dims.size(); ++i) {
+    if (transpose_dims[i] != i &&
+        !(absl::c_linear_search(lhs_contracting_dims, i) &&
+          absl::c_linear_search(lhs_contracting_dims, transpose_dims[i]))) {
+      return nullptr;
+    }
+  }
+  // Virtually pull the transpose into the dot. Now the dot is equivalent to
+  // a new dot with new lhs transpose dims equal to the transpose dims within
+  // the old lhs contracting dims.
+  auto old_lhs_contracting_dims = lhs_contracting_dims;
+  lhs_contracting_dims.Clear();
+  for (const auto dim : old_lhs_contracting_dims) {
+    lhs_contracting_dims.Add(lhs->dimensions(dim));
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // All checks are passed at this point.
+  //
+  // Transform lhs. Remove the transpose and reshape by sorting the lhs
+  // contracting dims and squish them into a single one. We don't actually
+  // squish the lhs_contracting_dims here because we still need the unsquished
+  // contracting dims to invert reshape and transpose.
+  absl::c_sort(lhs_contracting_dims);
+  lhs = computation_->AddInstruction(
+      HloInstruction::CreateReshape(reshape->shape(), lhs));
+
+  // Transform rhs.
+  // Let's walk through an example to see how rhs transformation is done and the
+  // meaning of some variables used in the implementation. Say we have the
+  // following original Hlo IR:
```

Comment:
Nit, we'd just say `s/Hlo IR/HLO/`.  That is, HLO is the name of the IR, "HLO IR" is redundant.